common/log.hpp

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_LOG_HPP
+#define MENDER_LOG_HPP
+
+#include "common_config.h"
+
+#include "loglevel.hpp"
+#include "logfield.hpp"
```

Comment:
I'm not gonna argue too much, but it's a little odd to me. Why would you ever change it out? And if you do, can't you then use `<>` ?

Reply:
Yes, you can change it later of course, but there's no good reason to use `""` in the first place. The main argument is that you can include a file whose name is conflicting with some other file. But this is not a good argument, because that file will then be impossible to include from a different folder. So any such conflicts should be resolved (rename, or include the name of the containing folder as a prefix, like Boost does) and we should rely on a correct include path from the build system instead.

Reply:
Yes, but this means you will always need the extra include in your compiler. 

I thought the main argument was that it shows your are including a local file?

Reply:
> Yes, but this means you will always need the extra include in your compiler.

Yes, and I think that's where it should be. The build system should be in charge of the include path, not the source files.

> I thought the main argument was that it shows your are including a local file?

This is true now, but may not be true later when we move things around. We may want to move internal header files into an `internal` folder, for example, which we should reflect in the include path, but not in each include statement.

We can also override certain headers this way, when we start splitting into platforms. In CFEngine [we had to change the style](https://github.com/cfengine/core/commit/66cb7d35b451100f60a3ba4197b7c7b46a45ae4d) to be able to build community and enterprise side by side, since some headers were different and therefore in different locations.

Reply:
I'd say the only exception is `#include "config.h"` if you have directory-specific `config.h` files instead of a single global one. And it's a good practice to have multiple ones.

Reply:
Even here I disagree, actually. 🙂 It's the build system's job to put the current directory first in the search path, if that's appropriate. Consider it like this: If we split source and header files completely into separate directories, which is quite common to do, then `""` is incorrect for all includes. On the other hand `<>` is never wrong as long as the build system is doing what it's supposed to.

Reply:
> It's the build system's job to put the current directory first in the search path

Agreed. And we should stick to reasonable build systems! :sweat_smile: 

To me `#include "config.h"` is nicer because it explicitly says where the `config.h` file is, but no strong opinion. And with a single global `config.h` using `#include <config.h>` is definitely the way to go.